### PR TITLE
std.algorithm.searching: minmaxElement

### PR DESCRIPTION
**Pull request by wilzbach**
follow-up to #4221 
An efficient combination of `{min,max}Element`.

Can be found in C++ as well:

http://en.cppreference.com/w/cpp/algorithm/minmax_element

Ping @nordlow.

### CODE REVIEW

File: std/algorithm/searching.d [hunk 1]
```diff
@@ -2960,6 +2960,162 @@ unittest
...
+body
+{
+    alias mapFun = unaryFun!map;
+    alias Element = ElementType!Range;
+    Unqual!Element seed = r.front;
+    r.popFront();
+    return minmaxElement!(map, selector)(r, seed);
+}
+
+private auto minmaxElement(alias map = "a", alias selector = "a < b", Range,
```
Comment (9il) [inline std/algorithm/searching.d]: 1. Why not to use common `map` instead?
2. Prototype with 2 seeds, one for max and one for min looks more useful
3. Why we need this function? is it faster then reduce!(min, max)?

Comment (9il) [inline std/algorithm/searching.d]: EDIT: 3. `map!(...).cache.reduce!(min, max)`

**Comment by wilzbach** [comment]
@9il - the idea comes from #4221 and @nordlow and I do prefer `.reduce(min, max)` too.
As mentioned in #4221 the problem is that having a `map` function will yield only the mapped minima.

If I think about it should be possible to templatize `extremum`, but on the other hand it will yield more complications and there is only `min`, `max`.

**Comment by 9il** [comment]
> @9il - the idea comes from #4221 and @nordlow and I do prefer .reduce(min, max) too.
> As mentioned in #4221 the problem is that having a map function will yield only the mapped minima.

I really don't understand. Could you please give an example where `minmaxElement` would be faster/better?

**Comment by DmitryOlshansky** [comment]
> Could you please give an example where minmaxElement would be faster/better?

It all comes down to having the original element not the one mapped by a predicate.
So that e.g. finding minimum a record by sub-field yields the record which has minimum sub-field not the minimum subfield value.

**Comment by wilzbach** [comment]
> I really don't understand. Could you please give an example where minmaxElement would be faster/better?

It will be faster than calling `extremum` twice ;-)
Btw for future readers as explained at #4221, it's not possible to get the element before `map` back, e.g.

```
assert([3, 4, 5, 1, 2].enumerate.minmaxElement!"a.value" ==
                                      tuple(tuple(3, 1), tuple(2, 5)));
```

vs.

```
[3, 4, 5, 1, 2].map!"a.value".reduce!(min, max) == tuple(1, 5);
```

**Comment by 9il** [comment]
Why not this:
```d
alias minValue = (a, b) => a.value < b.value ? a : b;
alias maxValue = (a, b) => a.value > b.value ? a : b;
assert([3, 4, 5, 1, 2].enumerate.reduce!(minValue, maxValue) ==  tuple(tuple(3, 1), tuple(2, 5)));
```
?

**Comment by wilzbach** [comment]
> Why not this:

Same reason as for #4221 applies (map function is evaluated twice), but let's focus the discussion on #4221.

**Comment by 9il** [comment]
could you please add prototype with 2 seeds?

**Comment by wilzbach** [comment]
> could you please add prototype with 2 seeds?

done - I removed the one seed prototype as it didn't make sense to me anymore.

**Comment by DmitryOlshansky** [comment]
Needs a changelog entry.

**Comment by wilzbach** [comment]
> Needs a changelog entry.

I know, already [noted](https://github.com/dlang/phobos/pull/4228) - I hope we can make the transition to the changelog folder soon ;-)

**Comment by wilzbach** [comment]
1) Please add @andralex label
2) Should I add the optimization for the noop map function case (=no mapping) to this function too?

File: std/algorithm/searching.d [hunk 2]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+                maxElementMapped = mapElement;
+            }
+            r.popFront();
+        }
+    }
+    return minmaxTuple(minElement, maxElement);
+}
+
+///
+//@safe pure unittest
```
Comment (JackStouffer) [inline std/algorithm/searching.d]: Fix this and the stdio import

File: std/algorithm/searching.d [hunk 3]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+    map = custom accessor for the comparison key
+    selector = custom mapping for the extrema selection
+    minSeed = custom seed to use as initial element for min
+    maxSeed = custom seed to use as initial element for min
+    r = Range from which the extreme value will be selected
+
+Returns:
+    The extreme value according to `map` and `selector` of the passed-in values.
+
+See_Also:
```
Comment (JackStouffer) [inline std/algorithm/searching.d]: Should also provide a link to the C++ version

**Comment by JackStouffer** [comment]
I think this function is of limited value. But, because there is an STL implementation adding this will help people transitioning from C++.

LGTM sans comments

File: std/algorithm/searching.d [hunk 4]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
     assert(arr2d.maxElement!"a[1]" == arr2d[1]);
 }
 
+/**
+Iterates the passed range and selects the min and max element with `less`.
+If the extreme element occurs multiple time, the first occurrence will be
+returned.
+This function is more efficient than calling both $(LREF minElement) and
+$(LREF maxElement).
```
Comment (JackStouffer) [inline std/algorithm/searching.d]: Could go into a little more detail here. I think you should add something like 

> This function is more efficient than calling both $(LREF minElement) and $(LREF maxElement) for one range because this function only requires one scan of the range, whereas the former takes two. Also, calling both $(LREF minElement) and $(LREF maxElement) on the same range would require it to be a forward range.

File: std/algorithm/searching.d [hunk 5]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+
+Params:
+    map = custom accessor for the comparison key
+    selector = custom mapping for the extrema selection
+    minSeed = custom seed to use as initial element for min
+    maxSeed = custom seed to use as initial element for min
+    r = Range from which the extreme value will be selected
+
+Returns:
+    The extreme value according to `map` and `selector` of the passed-in values.
```
Comment (JackStouffer) [inline std/algorithm/searching.d]: Should be something like

> A $(XREF Tuple, typecons) of the extreme values according to `map` and `selector` of the range.

File: std/algorithm/searching.d [hunk 6]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+If the extreme element occurs multiple time, the first occurrence will be
+returned.
+This function is more efficient than calling both $(LREF minElement) and
+$(LREF maxElement).
+
+Params:
+    map = custom accessor for the comparison key
+    selector = custom mapping for the extrema selection
+    minSeed = custom seed to use as initial element for min
+    maxSeed = custom seed to use as initial element for min
```
Comment (JackStouffer) [inline std/algorithm/searching.d]: Typo here, should be "custom seed to use as initial element for **max**"

File: std/algorithm/searching.d (hunk 3 above)
Comment (nordlow) [inline std/algorithm/searching.d]: I agree.

File: std/algorithm/searching.d [hunk 7]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+    alias CommonElement = CommonType!(Element, RangeElementType);
+    alias MapType = Unqual!(typeof(mapFun(CommonElement.init)));
+
+    Unqual!CommonElement minElement = minSeed;
+    Unqual!CommonElement maxElement = maxSeed;
+
+    MapType minElementMapped = mapFun(minElement);
+    MapType maxElementMapped = mapFun(maxElement);
+
+    alias minmaxTuple = Tuple!(Unqual!CommonElement, "min", Unqual!CommonElement, "max");
```
Comment (nordlow) [inline std/algorithm/searching.d]: Add unittests for access via members `min` and `max`. Otherwise LGTM.

**Comment by wilzbach** [comment]
> Add unittests for access via members min and max. Otherwise LGTM.

@nordlow this PR is a bit stalled, because I would like to get more feedback on #4265 (using just one loop statement is a lot faster) first. Most likely I might have to adapt this PR too then. Thanks for your feedback :)

File: std/algorithm/searching.d [hunk 8]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+    int[] arr;
+    assert(arr.minmaxElement(1, 2) == tuple(1, 2));
+}
+
+@safe pure unittest
+{
+    import std.range: enumerate, iota;
+    // supports mapping
+    assert([3, 4, 5, 1, 2].enumerate.minmaxElement!"a.value" ==
+                                    tuple(tuple(3, 1), tuple(2, 5)));
```
Comment (nordlow) [inline std/algorithm/searching.d]: Indent looks incorrect.

File: std/algorithm/searching.d [hunk 9]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+    assert([5, 2, 4].enumerate.minmaxElement!"a.value" ==
+                                    tuple(tuple(1, 2), tuple(0, 5)));
```
Comment (nordlow) [inline std/algorithm/searching.d]: ditto

File: std/algorithm/searching.d [hunk 10]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+
+    // forward ranges
+    assert(iota(1, 5).minmaxElement() == tuple(1, 4));
+    assert(iota(2, 5).enumerate.minmaxElement!"a.value" ==
+                                    tuple(tuple(0, 2), tuple(2, 4)));
```
Comment (nordlow) [inline std/algorithm/searching.d]: ditto

**Comment by JackStouffer** [comment]
Ping

File: std/algorithm/searching.d (hunk 3 above)
Comment (andralex) [inline std/algorithm/searching.d]: I agree, this won't stop decision to pull.

File: std/algorithm/searching.d [hunk 11]
```diff
@@ -3244,6 +3244,162 @@ auto maxElement(alias map = "a", Range, RangeElementType = ElementType!Range)
...
+    maxSeed = custom seed to use as initial element for min
+    r = Range from which the extreme value will be selected
+
+Returns:
+    The extreme value according to `map` and `selector` of the passed-in values.
+
+See_Also:
+    $(LREF minElement), $(LREF maxElement)
+*/
+auto minmaxElement(alias map = "a", alias selector = "a < b", Range)(Range r)
```
Comment (andralex) [inline std/algorithm/searching.d]: Can you define two overloads so there's no default for `map`?

**Comment by andralex** [comment]
I approve the name add but as we try to stay away from string lambdas and specializing on their values, please investigate using overloads instead of defaults for `map`. Thx!

**Comment by wilzbach** [comment]
> I approve the name add but as we try to stay away from string lambdas and specializing on their values, please investigate using overloads instead of defaults for map. Thx!

As said above I would prefer if we solve #4265 first, as it affects the design of this PR too ;-)

**Comment by wilzbach** [comment]
Just saw that there is [Boost.minmax](http://www.boost.org/doc/libs/1_62_0/libs/algorithm/minmax).

> As said above I would prefer if we solve #4265 first, as it affects the design of this PR too ;-)

Btw we are still blocked on this design decision :/

**Comment by wilzbach** [comment]
Oh and I bumped over an [old article](http://www.drdobbs.com/generic-min-and-max-redivivus/184403774) from @andralex about `min` in C++:

> Well I can't repress a smile when thinking of how people — or maybe myself, hopefully only a bit older and a lot wiser — will look at this article in 20 years. "Amazing, in 2001 these guys were having trouble implementing generically the min and max no-brainers in the most popular programming language of the time. Hah, it took this guy an entire article and some esoteric techniques to get min and max right."
> Maybe min and max aren't important? I'd argue to the contrary. Minimum and maximum are two simple concepts present both in math and in the real life. If a programming language is unable to express simple concepts of math and life, then there is something deeply wrong with that language.

So we still have five years to fix `min` in D :)

**Comment by andralex** [comment]
Thought min was fixed in D! What's wrong with it?

**Comment by wilzbach** [comment]
> Thought min was fixed in D! What's wrong with it?

Oh that was more intended as a joke, because there's no `minmax` yet ;-)

**Comment by nordlow** [comment]
Are we certain we want to return a tuple here and not a struct? I've noticed that use of `std.typecons : Tuple` introduces a significant compilation time overhead compare to structs. `findSplit*` uses Voldemort structs as returns.

**Comment by wilzbach** [comment]
> Are we certain we want to return a tuple here and not a struct? I've noticed that use of std.typecons : Tuple introduces a significant compilation time overhead compared to (Voldemort) structs.

Good point! This is still blocked to to the API question for [extremum](https://github.com/dlang/phobos/pull/4265#issuecomment-266234691). Unfortunately there's also a huge runtime difference between the "generic" version of `extremum` and a specialized one for the identity function. Condensed: the proposed API for `extremum` could either use (a) string comparisons `map == "a"` or (b) a lot more overloads.

**Comment by wilzbach** [comment]
Preview e.g. [here](http://dtest.thecybershadow.net/artifact/website-dec5e59a4d0eb4d0f2a298e9a3b5fcaf263316e9-ed7ce13923ee095e617c17026098b62e/web/library-prerelease/std/algorithm/comparison/among.html)

**Comment by andralex** [comment]
> the proposed API for extremum could either use (a) string comparisons map == "a" or (b) a lot more overloads.

isn't it only one more overload (without a mapping alias)?

**Comment by andralex** [comment]
If we need to add `minMaxElement`, we have lost. We have generic ways to compute multiple aggregate functions in one pass (`reduce`) so this specialization seems superfluous and a concession made to... what? convenience? efficiency? 

I suggest we take a step back and figure out what `minMaxElement` brings over `reduce(min, max)` and figure out whether we want it so bad and so frequently so as to add as its own thing.

What's next? `avgStddevElement`?

**Comment by andralex** [comment]
`sumAvgStddevElement`?

**Comment by andralex** [comment]
etc :)

**Comment by nordlow** [comment]
Maybe "we" just think that we need it because C++ added it...

**Comment by wilzbach** [comment]
> If we need to add minMaxElement, we have lost. We have generic ways to compute multiple aggregate functions in one pass (reduce) so this specialization seems superfluous and a concession made to... what? convenience? efficiency?

Efficiency was the motivation for C++ if I am not mistaken:

> Complexity: At most `max(floor(3/2(N−1)), 0)` applications of the predicate, where `N = std::distance(first, last)`.

I gave a benchmark a shot:

https://gist.github.com/wilzbach/425c72f028d87b19848c651e3d5a7063

Without a mapping function
--------------------------------------

- dmd is at least 10x slower than ldc
- interestingly `ldc` doesn't like the generic `minmaxElement` either

=> as expected with `ldc` there's _no_ measurable advantage between `reduce` and `minmaxElement`

With a mapping function
--------------------------------

This is the case where a custom mapping function `m` should be applied to each element, but one wants to obtain the original element.
The "naive" solution with reduce would be:

```d
return arr.reduce!(
   (a, b) => min(m(a), m(b)),
   (a, b) => max(m(a), m(b)),
);
```

We evaluate `m` twice as many times as needed and of course this doesn't lead to a good performance:

With `alias m = a => a + a`
----------------------

```
> ldc -release -O5 -mcpu=native test.d && ./test
reduce!(min,max) = 16 secs, 745 ms, 280 μs, and 1 hnsec
fold.minMax     = 13 secs, 756 ms, 979 μs, and 4 hnsecs
minmaxElement   = 14 secs, 179 ms, 390 μs, and 1 hnsec
```

Of course once the mapping function gets more expensive, the naive method takes much more time: 

With `alias m = a => a * a`
----------------------

```
> ldc -release -O5 -mcpu=native test.d && ./test
reduce!(min,max) = 25 secs, 262 ms, 508 μs, and 9 hnsecs
fold.minMax     = 15 secs, 106 ms, 617 μs, and 7 hnsecs
minmaxElement   = 14 secs, 807 ms, 857 μs, and 1 hnsec
```

What could the user do?
--------------------------------

So this is a simple solution that avoid the redundant invocation of the `map` function:

```d
static struct MinMaxTuple(T) {
    T min, max, minMap, maxMap;
    T opIndex(size_t i) { return i ? max : min; }
}
return arr[1..$].fold!((a, b) {
    auto c = m(b);
    if (c < a.minMap)
    {
        a.min = b;
        a.minMap = c;
    }
    if (c > a.maxMap)
    {
        a.max = b;
        a.maxMap = c;
    }
    return a;
})(MinMaxTuple!size_t(arr[0], arr[0], m(arr[0]), m(arr[0])));
```

For the source code and the DMD runtimes, see [the gist](https://gist.github.com/wilzbach/425c72f028d87b19848c651e3d5a7063).

So in summary
--------------------

If running `minmax` with a mapping function happens to be used often, than we should add it.
Another solution would be to expose `extremum` and let it accept multiple selector functions, s.t.  one can do `arr.extremum!(m, min, max)`

**Comment by andralex** [comment]
Ah, cool, I'd forgotten about the algorithmic trick. But I don't seem to see you applying it. The pattern goes:

```d
if (r[i] < r[i + 1])
{
    if (r[i] < r[min]) min = i;
    if (r[i + 1] > r[max]) max = i + 1;
}
else
{
    if (r[i + 1] < r[min]) min = i + 1;
    if (r[i] > r[max]) max = i;
}
```

It's actually one of the FB interview questions :). So yes I approve the addition, but for the life of me I don't see where you implement the correct algorithm above.

**Comment by wilzbach** [comment]
> Ah, cool, I'd forgotten about the algorithmic trick. But I don't seem to see you applying it. The pattern goes:

Yeah I tried it (see benchmarks below) and the "stupid" way seems to be a lot faster.
FWIW in C++ GCC and CLang also your pattern:

https://github.com/gcc-mirror/gcc/blob/gcc-7_1_0-release/libstdc%2B%2B-v3/include/bits/stl_algo.h#L3332
https://github.com/llvm-mirror/libcxx/blob/release_40/include/algorithm#L2662

I currently don't have time to dive more into it, but here are the benchmarks to compare the naive version and iteration in pairs:

RandomAccess
---------------------

```
ldc -release -O5 -mcpu=native test.d && ./test
reduce!(min,max) = 7 secs, 722 ms, 265 μs, and 4 hnsecs
fold.minMax     = 5 secs, 770 ms, and 427 μs
minmaxElement   = 5 secs, 410 ms, and 31 μs
minmaxElementInPairs = 15 secs, 362 ms, 724 μs, and 3 hnsecs
```

InputRange
---------------

```
>  ldc -release -O5 -mcpu=native test.d && ./test
reduce!(min,max) = 7 secs, 215 ms, 601 μs, and 1 hnsec
fold.minMax     = 6 secs, 507 ms, 278 μs, and 5 hnsecs
minmaxElement   = 6 secs, 361 ms, 248 μs, and 8 hnsecs
minmaxElementInPairs = 12 secs, 344 ms, 548 μs, and 3 hnsecs
```

([updated code from above](https://gist.github.com/wilzbach/3407d80bfa757d46a3ac59a873d5f085)).

**Comment by andralex** [comment]
Heh, thanks @wilzbach. I've reproed your measurements. This is an interesting result. I'm not sure exactly what's going on yet. Take a look at https://godbolt.org/g/mXj3uW. There we have:

* minmaxElementNoMap the brute force version. From what I can see ldc does some really awesome loop unrolling there,
* minmaxElementNoMap2 cleverer version that saves on comparisons under the assumption that if something is smaller than the smallest, it can't be greater than the greatest. So it does statistically between n and 2 * n comparisons. I measured no performance difference on any input. I ascribe this to the fact that the comparisons are done in parallel. LDC again generates interesting code, though very different from the first!
* minmaxElementNoMapInPairs the classic algorithm that does 3n/2 comparisons. It is indeed slower! LDC seems to generate much more conservative code.
* minmaxElementNoMapInPairs2 an improved version that has a tighter loop. It does improve the situation but not by much.

I'd say let's add minMaxElement with the guaranteed 3n/2 comparisons (which may be arbitrarily expensive); otherwise there is no merit to it over reduce!(min, max). We may specialize it for certain data types and the default comparison to take the brute force approach.

Compiler experts @ibuclaw  @JohanEngelen @klickverbot please take a look!

FWIW dmd generates equally good/bad code for minmaxElementNoMap2 and minmaxElementNoMapInPairs2. Didn't try gdc yet.

**Comment by wilzbach** [comment]
> @andralex: Do you have a benchmark script for your experiments?

@klickverbot: sorry that my link was so hidden at the end of the post. It's here:

https://gist.github.com/wilzbach/3407d80bfa757d46a3ac59a873d5f085

**Comment by andralex** [comment]
@klickverbot pasted the mess here: https://dpaste.dzfl.pl/09fbcf17f932

**Comment by JohanEngelen** [comment]
@andralex For the InPairs implementations, shouldn't the loop advance in pairs (+2)? (and then some extra work for odd length)

**Comment by andralex** [comment]
@JohanEngelen I ignored the odd elements case, it has no bearing on measuring efficiency. Per lines 124 and 202, I advance in 2 increments and look at i - 1 and i in one pass. If you find any bug I'd be relieved! I'm getting results that are difficult to interpret.

File: std/algorithm/searching.d (hunk 4 above)
Comment (andralex) [inline std/algorithm/searching.d]: More precisely, we should provide guarantees similar to the C++ version - per http://en.cppreference.com/w/cpp/algorithm/minmax_element: "At most max(floor(3/2(N−1)), 0) applications of the predicate, where N = std::distance(first, last)."

File: std/algorithm/searching.d (hunk 11 above)
Comment (andralex) [inline std/algorithm/searching.d]: Reiterating this request.

File: std/algorithm/searching.d [hunk 12]
```diff
@@ -3551,6 +3551,197 @@ if (isInputRange!Range && !isInfinite!Range &&
...
+/// ditto
+auto minmaxElement(alias map = "a", alias selector = "a < b", Range,
+                      RangeElementType = ElementType!Range)
+                     (Range r, RangeElementType minSeed, RangeElementType maxSeed)
+    if (isInputRange!Range && !isInfinite!Range &&
+        !is(CommonType!(ElementType!Range, RangeElementType) == void))
+{
+    alias mapFun = unaryFun!map;
+    alias selectorFun = binaryFun!selector;
+
```
Comment (andralex) [inline std/algorithm/searching.d]:
```d
assert(!selector(maxSeed, minSeed));
```

File: std/algorithm/searching.d [hunk 13]
```diff
@@ -3551,6 +3551,197 @@ if (isInputRange!Range && !isInfinite!Range &&
...
+            if (selectorFun(mapElement, minElementMapped))
+            {
+                minElement = r[i];
+                minElementMapped = mapElement;
+            }
+            if (selectorFun(maxElementMapped, mapElement))
+            {
+                maxElement = r[i];
+                maxElementMapped = mapElement;
+            }
```
Comment (andralex) [inline std/algorithm/searching.d]: So here we should use the 3n/2 algorithm, even if technically slower for `<` and `int`. One great thing to do would be to specialize for this (and a few other) cases.

File: std/algorithm/searching.d [hunk 14]
```diff
@@ -3551,6 +3551,197 @@ if (isInputRange!Range && !isInfinite!Range &&
...
+    }
+    else
+    {
+        while (!r.empty)
+        {
+            MapType rawElement1 = r.front;
+            MapType mapElement1 = mapFun(rawElement1);
+            r.popFront();
+            // check if the range had an uneven amount of elements and thus has ended
+            if (r.empty)
```
Comment (andralex) [inline std/algorithm/searching.d]: BUG: must return at the end of this if.

**Comment by andralex** [comment]
@wilzbach one more thing - where did you see the `getpid` trick in `doNotOptimizeAway`? I've seen it elsewhere, too, but forgot where. Thx!

**Comment by wilzbach** [comment]
> @wilzbach one more thing - where did you see the getpid trick in doNotOptimizeAway? I've seen it elsewhere, too, but forgot where. 

Ideally we get something like this trick into Phobos, s.t. users don't have to worry about it:

https://github.com/dlang/phobos/pull/5416

**Comment by JohanEngelen** [comment]
@andralex The code you linked to (https://github.com/dlang/phobos/pull/4248#issuecomment-306008191, https://godbolt.org/g/mXj3uW ) does not do the +2.

**Comment by JohanEngelen** [comment]
@andralex Did you trying "caching" `r.length` for code like this:
```d
for (size_t i = 0; i < r.length; i += 2)
{
    uint j = selectorFun(r[i], r[i + 1]);
```
in case the compiler cannot/doesnot deduce that `selectorFun` is not touching `r.length?

**Comment by andralex** [comment]
@JohanEngelen tried that, makes no difference.

BUT! I found what seems to be an interesting performance bug. I stripped minmaxElementInPairsNoMap all the way down to this core loop:

```d
    for (size_t i = 0; i < r.length; i += 2)
    {
    }
```

Even if it literally does nothing, it still takes more than 2 times longer than `minmaxElementNoMap`. If I change it to:

```d
    for (size_t i = 0; i < r.length; i += 1)
    {
    }
```

So ldc does not generate good code for loops that advance in a non-unit increment. I think you'd improve the life of many if you looked into that!

**Comment by JohanEngelen** [comment]
@andralex That's because of integer overflow. If `r.length == size_t.max` (an odd number!), it's an infinite loop.
See https://godbolt.org/g/5RFNdt  . With the early `r.length == size_t.max` return path, GDC is able to optimize all out. LDC isn't.  (I filed https://github.com/ldc-developers/ldc/issues/2154 )

**Comment by andralex** [comment]
@JohanEngelen fwiw I changed the implementation to use `++i` for iteration up to `r.length / 2` and used `2 * i` and `2 * i + 1` as adjacent elements. Still no improvement.

@wilzbach so let's stay with the `3n/2` algo for this PR. Works?

**Comment by wilzbach** [comment]
I never needed this and lost interest in pursueing this PR. Sorry
